### PR TITLE
[REF] Fix ORM cyclic imports and move to odoo.orm package

### PR DESCRIPTION
**Pull request by kmagusiak**
Description of the issue/feature this PR addresses:
Move to `odoo.orm` package: api, fields, models, registry. These contain the main parts of the ORM

Current behavior before PR:
Imports done at the end of the files because of dependencies during imports. Core files are getting big (mostly `fields.py` at ~200kb).

Desired behavior after PR is merged:
Regroup ORM files in a single location, split big files and fix import loops.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/api/__init__.py
Comment (Julien00859) [inline odoo/api/__init__.py:1]: Add a comment explaining why this is `api/__init__.py` and not `api.py`: to ease forward-ports. Do the same for the other modules. Maybe add a "TODO Odoo 22: rename `api/__init__.py` => `api.py`"

File: odoo/orm/__init__.py
Comment (Julien00859) [inline odoo/orm/__init__.py:1]: maybe rename to `_orm`, so it is clear that it is our internal stuff, that people should keep on importing from models/fields/api and not orm

**Comment by kmagusiak** [comment]
Previously existing files are in a folder with a single `__init__.py` file because:
- This way the old file does not exist, which allows git to follow the rename and better resolve conflicts.
- We don't end with the same file name twice in the ORM: searching for `fields.py` would return only the one in the new module.

File: odoo/api/__init__.py
Comment (rco-odoo) [inline odoo/api/__init__.py:23]: Some of those types were always available.  Shouldn't we remain backward compatible?
```suggestion
if typing.TYPE_CHECKING:
    from odoo._orm.types import BaseModel, IdType, NewId, Self
from odoo._orm.types import ContextType, DomainType, ValuesType
```

File: odoo/addons/test_new_api/tests/test_new_fields.py
Comment (rco-odoo) [inline odoo/addons/test_new_api/tests/test_new_fields.py:4401]: Meh.  That namespace sucks.
We should either use `odoo.fields` for all field-specific stuff, or use prefix `odoo.orm` for all ORM-specific stuff (`odoo.orm.models`, `odoo.orm.fields`, etc.)

File: odoo/tools/config.py
Comment (rco-odoo) [inline odoo/tools/config.py:214]: Meh, as above.

File: odoo/_orm/__init__.py
Comment (kmagusiak) [reply odoo/_orm/__init__.py]: Split done. In a second time we can move things to `tools` and `util`. Other functions from `api` were not used elsewhere, that is why I did not move them.

File: odoo/addons/test_new_api/tests/test_new_fields.py
Comment (kmagusiak) [reply odoo/addons/test_new_api/tests/test_new_fields.py]: I have aligned the loggers in all the files to `odoo.orm.*`.

File: odoo/addons/base/tests/test_ormcache.py
Comment (kmagusiak) [reply odoo/addons/base/tests/test_ormcache.py]: See next comments for the loggers.

File: odoo/api/__init__.py
Comment (kmagusiak) [reply odoo/api/__init__.py:23]: No, because these are introduced only for type-checking, importing `odoo._orm.types` without typing context will lead to cycles in dependencies.

Comment (kmagusiak) [reply odoo/api/__init__.py]: Exposed as before.

**Review by ryv-odoo — APPROVED** [review]

File: odoo/_orm/fields.py
Comment (rco-odoo) [inline odoo/_orm/fields.py:27]: You actually don't need the `TYPE_CHECKING` case.  You could do the `else` clause all the time in this file.

File: odoo/_orm/identifiers.py
Comment (rco-odoo) [inline odoo/_orm/identifiers.py:41]: Move this line to `types`?

File: odoo/_orm/relational_fields.py
Comment (rco-odoo) [inline odoo/_orm/relational_fields.py:1308]: Duplicate? :smile:

File: odoo/addons/base/tests/test_ormcache.py
Comment (rco-odoo) [inline odoo/addons/base/tests/test_ormcache.py:26]: Why?

File: odoo/_orm/fields.py
Comment (kmagusiak) [reply odoo/_orm/fields.py]: Removed.

File: odoo/_orm/identifiers.py
Comment (kmagusiak) [reply odoo/_orm/identifiers.py:41]: It must be here, because it is needed in basic fields to declare `Id` and `types` cannot be imported outside a type-checking block in the ORM without causing too many dependency loops.

File: odoo/addons/base/tests/test_ormcache.py
Comment (kmagusiak) [reply odoo/addons/base/tests/test_ormcache.py]: Just replacement with the logs caught this one too. Fixed.

**Comment by rco-odoo** [comment]
Can we put this pull request on pause for a while?  This still doesn't look right to me.  I think it deserves some extra maturing.

To summarize, I think that it fixes two issues:
- circular imports in the ORM's implementation, which could potentially be fixed in a simpler way (for instance, the import of fields in `models.py` may entirely be replaced by local imports);
- accidental leakage of the ORM's implementation (stripping down the module's API would be better).

The proposed file structure for the ORM's implementation does not really look clean to me.  There are too many hacks like `_Fields_BaseModel` that aren't straightforward and even feel bug-prone.

**Comment by Xavier-Do** [comment]
I'm not following this pr deeply but is still would like to give a general opinion.

image

ORM core files contains a rich history than can be useful in some situations

Today, when browsing fields.py and model.py, I can easily know if a change is recent or not just looking at the info given by git lens and heatmap. It is a precious information when investigating a change of behavior, quickly locate a change that was made recently. By recently I mean, in the last 1-2 years, compared to code modified 4-10 years ago. With this version 90% of this history will be "modified a few months by krma". Anytime a line of code is moved from one file to another, it is 10 times more painful to find the orign of this line of code.

I don't say that this is a reason valid enough to avoid a refactoring if needed, refactoring the code is necessary to keep something clean in the long run that will allow odoo to survive for a long time, but once it is merged, this ease to blame old code is lost.  So my question is the following: will this change give sufficient benefit so that it worst loosing the immediate access to the git history. 

Removing circular import looks like a good idea if done correctly, it may solve some issues regarding the initial code organisation, remove some hacks maybe. But splitting a big file in smaller part here has low value  IMHO.

**Comment by Xavier-Do** [comment]
You example shows one file where a rename was detected but it is not the case for all other files... 

Before
image
After
image

Moreover when switching from one version to another all the fields opened in the editor are not automatically reopened so even is git knows the move is inconvenient. 

Again, it is not a review saying that you cannot merge it but just a personal opinion, I know that blaming  will be more frustrating in the future because of this. I needed to at least signal that. If other reviewer think that it is needed, they will merge it anyway.

**Comment by kmagusiak** [comment]
> Again, it is not a review saying that you cannot merge it but just a personal opinion, I know that blaming will be more frustrating in the future because of this. I needed to at least signal that. If other reviewer think that it is needed, they will merge it anyway.

We will have a lot of new features introduced in coming months, that's why I suggest to do it now (or not at all).
If we split `fields.py`, I think this is the best that can be done.

**Review by rco-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@robodoo override=ci/security

Code has been moved almost as-is.  All the spotted cases were already present and known.

**Comment by rco-odoo** [comment]
@robodoo override=ci/style

We'll fix that later.  The purpose of this PR is to move code, not fix its style issues.

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

File: odoo/orm/types.py
Comment (Xavier-Do) [reply odoo/orm/types.py:13]: Master is 3.10 and will remain 3.10 for a while.

Please fix this asap ;)

Comment (kmagusiak) [reply odoo/orm/types.py:13]: #184530 for the fix

**Pull request closed without merging**
